src/backend/distributed/commands/multi_copy.c

```diff
@@ -465,7 +465,7 @@ CopyToExistingShards(CopyStmt *copyStatement, char *completionTag)
 	 * PG_CATCH. Otherwise, it may be undefined in the PG_CATCH (see sigsetjmp
 	 * documentation).
 	 */
-	shardConnectionHash = CreateShardConnectionHash();
+	copyConnectionHash = CreateShardConnectionHash(MessageContext);
```

Comment:
On 2016-09-02 09:35:03 -0700, Jason Petersen wrote:
> > @@ -465,7 +465,7 @@ CopyToExistingShards(CopyStmt *copyStatement, char *completionTag)
> >  	 * PG_CATCH. Otherwise, it may be undefined in the PG_CATCH (see sigsetjmp
> >  	 * documentation).
> >  	 */
> > -	shardConnectionHash = CreateShardConnectionHash();
> > +	copyConnectionHash = CreateShardConnectionHash(MessageContext);
> 
> Previously, no context was passed, but the context flag was set. This results in a brand-new parentless memory context for the hash to use for allocations, i.e. a memory leak.

FWIW, I think we should backpatch that, i.e. fix it individually.

> Simply removing the context flag would be wrong, as that just uses `TopMemoryContext` as the parent instead (not much better).
> 
> I figured `MessageContext` would make sense, as this hash only lives for one message (the `COPY`), right?

I'm generally loathe to use MessageContext for longer lived stuff. It's
not likely to be a problem here, but MessageContext really can be reset
by protocol messages. That could e.g. become a problem if a COPY is done
from a set returning function, which is queried via a cursor.